src/entities.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+"""Generates entities.h and entities.c using entities.json from https://html.spec.whatwg.org/entities.json
+
+Delete entities.json if you want to fetch the latest version.
+The current version has 1 and 2 codepoint entities.
+If the spec is updated to include 3 or more codepoints per entity, then this script along with
+gdImageStringFTEx and gd_Entity_To_Unicode (from gdft.c) will need to be modified.
+A warning is issued for any such entities found and only the first 2 codepoints are used.
+"""
+
+import json
+import os
+import os.path
+import re
+import urllib.request
+from pathlib import Path
+
+__author__ = "Sterling Pickens"
+__copyright__ = "Copyright 2021, The libgd Project"
+__credits__ = ["Sterling Pickens", "Mike Frysinger"]
+__license__ = "GPLv3 or later"
+__version__ = "1.0.0"
+__maintainer__ = "Sterling Pickens"
+__email__ = "sterling_pickens@yahoo.com"
+__status__ = "Production"
+
+# Ensure we are only in src dir
+SRCDIR = Path(__file__).resolve().parent
+os.chdir(SRCDIR)
+
+# Declare File content strings
+C_FILE_HEAD = """ /*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+ */
+
+#include "entities.h"
+
+const struct entities_s entities[NR_OF_ENTITIES] = {
+"""
+H_FILE_HEAD = """ /*
+ * Generated file - do not edit directly.
+ * This file was generated from:
+ *     https://html.spec.whatwg.org/entities.json
+ * by means of entities.py in the libgd src dir
+ */
+
+#ifndef GD_ENTITIES_H
+#define GD_ENTITIES_H 1
+
+#include <stdlib.h>
+#include <stdint.h>
+
+"""
+H_FILE_TAIL = """
+#define ENTITY_HEX_LENGTH_MAX 10
+#define ENTITY_DEC_LENGTH_MAX 10
+
+// html entity strings are entity prefix + string + suffix limited
+// hex and dec should be limited to current unicode spec + entity prefix + suffix
+
+struct entities_s {
+                  	char *name;
+                        uint32_t unicode_a;
+                        uint32_t unicode_b;
+                };
+extern const struct entities_s entities[NR_OF_ENTITIES];
+
+#endif
+"""
+
+# Check for entities.json and fetch if needed
+if not os.path.exists("entities.json"):
+    print("entities.json not found, attempting to download...")
+    with open("entities.json", "wb") as file_json:
+        file_json.write(urllib.request.urlopen("https://html.spec.whatwg.org/entities.json").read())
+        print("\tsuccessful")
+
+# Load json obj
+with open("entities.json", "rb") as file_json:
+    entities = json.load(file_json)
+
+# Initialize some objects
+curline = 1
+name_matcher = re.compile(r"&\S+;")
+
+# Sum of total matching entities
+total_entities = sum(1 if name_matcher.match(key) else 0 for key in entities)
+
+# Find longest entity
+len_name_max = max(len(key) if name_matcher.match(key) else 0 for key in entities)
```

Comment:
this one we'd shuffle so that the generator doesn't produce anything if it doesn't match

`len_name_max = max(len(key) for key in entities if name_matcher.match(key))`